salt/modules/pkgng.py

```diff
@@ -1867,6 +1877,502 @@ def updating(name,
+        salt '*' pkg.lock <package name>
+
+    jail
+        Lock packages within the specified jail
+
+        CLI Example:
+
+        .. code-block:: bash
+
+            salt '*' pkg.lock <package name> jail=<jail name or id>
+
+    chroot
+        Lock packages within the specified chroot (ignored if ``jail`` is
+        specified)
+
+        CLI Example:
+
+        .. code-block:: bash
+
+            salt '*' pkg.lock <package name> chroot=/path/to/chroot
+
+    root
+        Lock packages within the specified root (ignored if ``jail`` is
+        specified)
+
+        CLI Example:
+
+        .. code-block:: bash
+
+            salt '*' pkg.lock <package name> root=/path/to/chroot
+
+    '''
+    if name in _lockcmd('lock', name, **kwargs):
+        return True
+
+    return False
+
+
+def unlock(name, **kwargs):
+    '''
+    Unlock the named package against reinstallation, modification or deletion.
+
+    Returns True if the named package was successfully unlocked.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' pkg.unlock <package name>
+
+    jail
+        Unlock packages within the specified jail
+
+        CLI Example:
+
+        .. code-block:: bash
+
+            salt '*' pkg.unlock <package name> jail=<jail name or id>
+
+    chroot
+        Unlock packages within the specified chroot (ignored if ``jail`` is
+        specified)
+
+        CLI Example:
+
+        .. code-block:: bash
+
+            salt '*' pkg.unlock <package name> chroot=/path/to/chroot
+
+    root
+        Unlock packages within the specified root (ignored if ``jail`` is
+        specified)
+
+        CLI Example:
+
+        .. code-block:: bash
+
+            salt '*' pkg.unlock <package name> root=/path/to/chroot
+
+    '''
+    if name in _lockcmd('unlock', name, **kwargs):
+        return False
+
+    return True
+
+
+def _lockcmd(subcmd, pkgname=None, **kwargs):
+    '''
+    Helper function for lock and unlock commands, because their syntax is identical.
+
+    Run the lock/unlock command, and return a list of locked packages
+    '''
+
+    jail = kwargs.pop('jail', None)
+    chroot = kwargs.pop('chroot', None)
+    root = kwargs.pop('root', None)
+
+    locked_pkgs = []
+
+    cmd = _pkg(jail, chroot, root)
+    cmd.append(subcmd)
+    cmd.append('-y')
+    cmd.append('--quiet')
+    cmd.append('--show-locked')
+
+    if pkgname:
+        cmd.append(pkgname)
+
+    out = __salt__['cmd.run_all'](cmd, output_loglevel='trace', python_shell=False)
+
+    if out['retcode'] != 0:
+        raise CommandExecutionError(
+            'Problem encountered {0}ing packages'.format(subcmd),
+            info={'result': out}
+        )
+
+    for line in salt.utils.itertools.split(out['stdout'], '\n'):
+        if not line:
+            continue
+        try:
+            pkgname = line.rsplit('-', 1)[0]
+        except ValueError:
+            continue
+        locked_pkgs.append(pkgname)
+
+    log.debug("Locked packages: {0}".format(locked_pkgs))
```

Comment:
Can you revise your logging calls, and remove str.format usage in favor of string replacement? Python logging already does string replacement using any positional args passed after the format string, so using str.format is redundant. Also, we have code in our custom logging handler which normalizes the format string and the args passed to it so that they are unicode, preventing log messages from triggering unicode encoding/decoding errors. This log message should be changed to:

```python
    log.debug("Locked packages: %s", locked_pkgs)
```

I realize you have probably seen a lot of usage of str.format in logging, but it's something we're working to stamp out in the codebase to make logging work better with unicode.

Reply:
Note, there are other logging calls where similar changes need to be made, this is just one example.